File: Compiler/JackAnalyzer/CompilationEngine.py
from typing import NoReturn, TextIO
from exceptions import JackSyntaxError
from JackTokenizer import JackTokenizer
from GrammarType import *
from Keywords import *
from TokenType import TokenType

# test


def grammar_rule(rule):
    """Ensures the first and last thing that happens when a grammar rule is compiled is pushing and popping the rule from the stack. Call the decorator with the grammar rule as it's argument. """
    def decorator(fn):
        def wrapper(self):
            self.push_rule(rule)
            fn(self)
            self.pop_rule()
        return wrapper
    return decorator


class CompliationEngine:
    """
    Building a compiler for Jack language in the nand2tetris course.
    Note to self - All the helper methods are needed to avoid duplicate xml tags 
                   on a recursive call. This could be fixed by adding the xml tag
                   outside of the fuction itself. Add <expression> right before
                   a call to compile_expr(), rather than at the beginning of compile_expr().
    """

    def __init__(self, input: JackTokenizer, outfile: TextIO) -> None:
        self.input = input
        self.outfile = outfile
        self.rule: str = ''  # current grammar rule we are compiling
        self.rule_stack: List[str] = []
        self.token: str = ''
        self.token_type: TokenType = TokenType.NONE

        self.advance()

    @grammar_rule(CLASS_GRAMMAR)
    def compile_class(self) -> None:
        """
        'class' className '{' classVarDec* subroutineDec* '}'
        """
        self.eat(CLASS_KEYWORD)
        print(f'Compiling `{self.token}` class...')
        self.eat(TokenType.IDENTIFIER)
        self.eat('{')
        while self.token in (STATIC, FIELD):
            self.compile_classVarDec()
        while self.token != '}':
            self.compile_subroutineDec()
        self.eat('}')
        print('class compilation finished.')

    @grammar_rule(CLASS_VAR_DEC)
    def compile_classVarDec(self) -> None:  # sourcery skip: class-extract-method
        """
        Compiles a static or field variable declaration.
        Grammar:
         - ('static'|'field' ) type varName (',' varName)* ';'
        """

        self.eat(STATIC, FIELD)
        self.eat(*var_types)
        self.compile_varDecList()
        self.eat(';')

    @grammar_rule(SUBROUTINE_DEC)
    def compile_subroutineDec(self) -> None:
        """
        grammar: ('constructor'|'function'|'method') (void' | type) subroutineName
                 '(' parameterList ')' subroutineBody
        """
        self.eat(CONSTRUCTOR, FUNCTION, METHOD)
        self.eat(VOID, *var_types)
        print(f'Compiling `{self.token}` subroutine... ')
        self.eat(TokenType.IDENTIFIER)
        self.eat('(')
        self.compile_paramList()
        self.eat(')')
        self.compile_subroutineBody()
        print('subroutine compilation finished.')

    @grammar_rule(PARAMETER_LIST)
    def compile_paramList(self) -> None:
        """
        grammer: ( (type varName) (',' type varName)*)?
        """
        # no parameter list
        if self.token == ')':
            return

        self.eat(*var_types)
        self.eat(TokenType.IDENTIFIER)

        if self.token == ',':
            self.paramList_helper()

    def paramList_helper(self) -> None:
        self.eat(',')
        self.eat(*var_types)
        self.eat(TokenType.IDENTIFIER)
        if self.token == ',':
            self.paramList_helper()

    @grammar_rule(SUBROUTINE_BODY)
    def compile_subroutineBody(self) -> None:
        """
        grammar: '{' varDec* statements '}'
        """
        self.eat('{')
        while self.token == VAR:
            self.compile_varDec()

        self.compile_statements()

        self.eat('}')

    @grammar_rule(VAR_DEC)
    def compile_varDec(self) -> None:
        """
        Compiles a variable declaration.
        Grammar:
         - 'var' type varName (',' varName)* ';'
        """
        self.eat(VAR)
        self.eat(*var_types, TokenType.IDENTIFIER)
        self.compile_varDecList()
        self.eat(';')

    def compile_varDecList(self) -> None:
        self.eat(TokenType.IDENTIFIER)

        if self.maybe_eat(','):
            self.compile_varDecList()

    @grammar_rule(STATEMENTS)
    def compile_statements(self) -> None:
        while self.token in [LET, IF, DO, WHILE, RETURN]:
            method = getattr(self, f'compile_{self.token}')
            method()

    @grammar_rule(LET_STATEMENT)
    def compile_let(self) -> None:
        self.eat(LET)
        self.eat(TokenType.IDENTIFIER)

        if self.maybe_eat('['):
            self.compile_expr()
            self.eat(']')

        self.eat('=')
        self.compile_expr()
        self.eat(';')

    @grammar_rule(IF_STATEMENT)
    def compile_if(self) -> None:
        """
        grammar:
        'if' '(' expression ')' '{' statements '}' ('else' '{' statements '}')?
        """
        self.eat(IF)
        self.eat('(')
        self.compile_expr()
        self.eat(')')

        self.eat('{')
        self.compile_statements()
        self.eat('}')

        if self.token != ELSE:
            return

        # else block
        self.eat(ELSE)
        self.eat('{')
        self.compile_statements()
        self.eat('}')

    @grammar_rule(WHILE_STATEMENT)
    def compile_while(self) -> None:
        """
        'while' '(' expression ')''{' statements '}'
        """
        self.eat(WHILE)
        self.eat('(')
        self.compile_expr()
        self.eat(')')

        self.eat('{')
        self.compile_statements()
        self.eat('}')

    @grammar_rule(DO_STATEMENT)
    def compile_do(self) -> None:
        """
        'do' subroutineCall';'
        """
        self.eat(DO)
        self.eat(TokenType.IDENTIFIER)
        self.compile_subroutineCall()
        self.eat(';')

    @grammar_rule(RETURN_STATEMENT)
    def compile_return(self) -> None:
        """'return' expression? ';'"""
        self.eat(RETURN)
        if self.token != ';':
            self.compile_expr()
        self.eat(';')

    @grammar_rule(EXPRESSION)
    def compile_expr(self) -> None:
        """
        term (op term)*
        """
        self.expr_helper()

    def expr_helper(self) -> None:
        self.compile_term()
        if self.maybe_eat(*binary_ops):
            self.expr_helper()

    @grammar_rule(TERM)
    def compile_term(self) -> None:
        """
        integerConstant 
        | stringConstant 
        | keywordConstant 
        | varName 
        | varName '[' expression ']' 
        | subroutineCall 
        | '(' expression ') ' 
        | unary Op term

        """
        if self.maybe_eat(TokenType.INT_CONST, TokenType.STR_CONST, *keyword_constants):
            return
        elif self.maybe_eat('('):
            self.compile_expr()
            self.eat(')')
        elif self.maybe_eat(*unary_ops):
            self.compile_term()
        else:
            self.eat(TokenType.IDENTIFIER)
            if self.maybe_eat('['):
                self.compile_expr()
                self.eat(']')
            elif self.token in ('.', '('):
                self.compile_subroutineCall()

            # otherwise it's just a plain varName

    def compile_subroutineCall(self) -> None:
        """
        subroutineName '(' expressionList ')' 
        | (className| varName) '.' subroutineName '(' expressionList ')'        
        """
        if self.maybe_eat('.'):
            self.eat(TokenType.IDENTIFIER)

        self.eat('(')
        self.compile_exprList()
        self.eat(')')

    @grammar_rule(EXPRESSIONLIST)
    def compile_exprList(self) -> None:
        """(expression (',' expression)* )?"""
        self.exprList_helper()

    def exprList_helper(self) -> None:
        if self.token == ')':
            return

        self.compile_expr()
        if self.maybe_eat(','):
            self.exprList_helper()

    def eat(self, *args: str | TokenType) -> None:
        """
        VERY IMPORTANT FUNCTION
        It checks that the token or token_type is in one of the arguments, 
        throwing an error if it isn't, and advancing to the next token if it is. 
        """
        found = any(
            self.token == arg or self.token_type == arg
            for arg in args
        )

        if not found:
            self.error(f'{args}')

        self.write_token()
        self.advance()

    def maybe_eat(self, *args: str | TokenType) -> bool:
        """
        Same as `eat()` but it doesn't throw an error if none 
        of the arguments match self.token or self.token_type

        returns true if one of the arguments was found, false otherwise.
        """
        found = any(
            self.token == arg or self.token_type == arg
            for arg in args
        )

        if found:
            self.write_token()
            self.advance()

        return found

    def write_token(self):
        self.writeln(self.input.to_xml())

    def writeln(self, *lines: str):
        for line in lines:
            self.outfile.write(line + '\n')

    def error(self, expected: str) -> NoReturn:
        raise JackSyntaxError(
            f'\nERROR: Invalid {self.rule} \nExpected {expected} \nCurrent Token: {self.token} \nCurrent line: {self.input.get_line()} ')

    def advance(self):
        if self.input.has_more_tokens():
            self.input.advance()
            self.token = self.input.token_value()
            self.token_type = self.input.token_type()
        else:
            print('>>>WARNING: self.advance() was called when the analyzer had no more tokens.')

    def push_rule(self, rule: str) -> None:
        """pushes `rule` onto the rule_stack"""
        self.rule_stack.append(rule)
        self.rule = rule
        self.tag()

    def pop_rule(self) -> None:
        self.end_tag()
        rule = self.rule_stack.pop()
        self.rule = self.rule_stack[-1] if self.rule_stack else ''

    def tag(self) -> None:
        self.writeln(f'<{self.rule}>')

    def end_tag(self) -> None:
        self.writeln(f'</{self.rule}>')


class Dummy:
    def __init__(self) -> None:
        self.x = '1'

    def amethod(self) -> None:
        print('amethod')


if __name__ == '__main__':
    dummy = Dummy()
    getattr(dummy, 'amethod')()
    # attr()

File: CompilerInC/CompilationEngine.c
#include "CompilationEngine.h"
#include "JackTokenizer.h"
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

void eatOneOf(char **, size_t);
bool maybeEatOneOf(char **, size_t);
void eatTT(TokenType);
bool maybeEatTT(TokenType);
void eat(char *);
void eatReturnType(); // 'void' || eatType()
bool maybeEat(char *);
void eatType(); // int, char, boolean or className (identifier)
bool cmpTok(char *);
void advanceTokenizer();
void writeToken();
void startRule(GrammarRule);
void endRule(GrammarRule);
char *join(char **, char *, size_t);
void error(char *msg);
void compileSubCall();

FILE *out;
char *currentToken;
TokenType currentTokenType;

char *kwConstants[4] = {"true", "false", "null", "this"};
char *binOps = "+-*/&|<>=";
char *unaryOps = "-~";

const int MAX_TOKEN_SIZE = 255;
const int MAX_TOKEN_TYPE_SIZE = 16; // longest string retuned by ttStr()

void initCompilationEngine(FILE *_outFile) {
    out = _outFile;
    advanceTokenizer();
}

void compileClass() {
    /*
    'class' className '{' classVarDec* subroutineDec* '}'
    */
    startRule(CLASS_RULE);

    eat(kwStr(CLASS));
    eatTT(IDENTIFIER);
    eat("{");
    while (cmpTok(kwStr(STATIC)) || cmpTok(kwStr(FIELD)))
        compileClassVarDec();

    while (cmpTok(kwStr(CONSTRUCTOR)) || cmpTok(kwStr(METHOD)) || cmpTok(kwStr(FUNCTION)))
        compileSubroutineDec();

    eat("}");

    endRule(CLASS_RULE);
}


void compileClassVarDec() {
    /*
    ('static'|'field' ) type varName (',' varName)* ';'
    */
    startRule(CLASS_VAR_DEC);

    char *expect[2] = {kwStr(STATIC), kwStr(FIELD)};
    eatOneOf((char **)expect, 2);
    eatType();
    eatTT(IDENTIFIER);
    while (maybeEat(",")) {
        eatTT(IDENTIFIER);
    }
    eat(";");

    endRule(CLASS_VAR_DEC);
}

void compileSubroutineDec() {
    /*
    ('constructor' | 'function' | 'method') ('void' | type) subroutineName
    '(' parameterList ')' subroutineBody
    */
    startRule(SUBROUTINE_DEC);

    char *expect[3] = {kwStr(CONSTRUCTOR), kwStr(FUNCTION), kwStr(METHOD)};
    eatOneOf((char **)expect, 3);
    eatReturnType();
    eatTT(IDENTIFIER);
    eat("(");
    compileParamList();
    eat(")");
    compileSubroutineBody();
    endRule(SUBROUTINE_DEC);
}
void compileParamList() {
    /*
    ((type varName) (',' type varName)*)?
    */

    startRule(PARAMETER_LIST);

    if (!strncmp(currentToken, ")", 2)) {
        endRule(PARAMETER_LIST);
        return;
    }

    eatType();
    eatTT(IDENTIFIER);
    while (maybeEat(",")) {
        eatType();
        eatTT(IDENTIFIER);
    }

    endRule(PARAMETER_LIST);
}
void compileSubroutineBody() {
    /*
    '{' varDec* statements '}'
    */
    startRule(SUBROUTINE_BODY);
    eat("{");
    while (!strncmp(currentToken, kwStr(VAR), 4))
        compileVarDec();
    compileStatements();
    eat("}");
    endRule(SUBROUTINE_BODY);
}
void compileVarDec() {
    /*
    'var' type varName (', ' varName)* ';'
    */
    startRule(VAR_DEC);

    eat(kwStr(VAR));
    eatType();
    eatTT(IDENTIFIER);
    while (maybeEat(","))
        eatTT(IDENTIFIER);

    eat(";");

    endRule(VAR_DEC);
}
void compileStatements() {
    /*
    Statement*
    Statement:
    letStatement | ifStatement| whileStatement | doStatement | returnStatement
    */
    startRule(STATEMENTS);
    while (cmpTok(kwStr(LET)) || cmpTok(kwStr(IF)) || cmpTok(kwStr(WHILE)) || cmpTok(kwStr(DO))
           || cmpTok(kwStr(RETURN))) {
        if (cmpTok(kwStr(LET)))
            compileLet();
        else if (cmpTok(kwStr(IF)))
            compileIf();
        else if (cmpTok(kwStr(WHILE)))
            compileWhile();
        else if (cmpTok(kwStr(DO)))
            compileDo();
        else if (cmpTok(kwStr(RETURN)))
            compileReturn();
    }
    endRule(STATEMENTS);
}
void compileLet() {
    /*
    'let' varName ('[' expression ']')? '=' expression ';'
    */
    startRule(LET_STATEMENT);
    eat(kwStr(LET));
    eatTT(IDENTIFIER);
    if (maybeEat("[")) {
        compileExpr();
        eat("]");
    }
    eat("=");
    compileExpr();
    eat(";");
    endRule(LET_STATEMENT);
}
void compileIf() {
    /*

    'if' '(' expression ')' '{' statements "}' ('else' '{' statements '}')?
    */
    startRule(IF_STATEMENT);
    eat(kwStr(IF));
    eat("(");
    compileExpr();
    eat(")");

    eat("{");
    compileStatements();
    eat("}");
    if (maybeEat(kwStr(ELSE))) {
        eat("{");
        compileStatements();
        eat("}");
    }
    endRule(IF_STATEMENT);
}
void compileDo() {
    /*
    'do' subroutineCall';'
    */
    startRule(DO_STATEMENT);
    eat(kwStr(DO));
    eatTT(IDENTIFIER);
    if (maybeEat("."))
        eatTT(IDENTIFIER);
    eat("(");
    compileExprList();
    eat(")");
    eat(";");
    endRule(DO_STATEMENT);
}
void compileWhile() {
    /*

    'while' '(' expression ')' '{' statements '}'
    */
    startRule(WHILE_STATEMENT);

    eat(kwStr(WHILE));
    eat("(");
    compileExpr();
    eat(")");

    eat("{");
    compileStatements();
    eat("}");

    endRule(WHILE_STATEMENT);
}
void compileReturn() {
    /*

    'return' expression? ':'
    */
    startRule(RETURN_STATEMENT);

    eat(kwStr(RETURN));
    if (!maybeEat(";")) {
        compileExpr();
        eat(";");
    }

    endRule(RETURN_STATEMENT);
}
void compileExpr() {
    /*
    term (op term)*
    */
    startRule(EXPRESSION);
    compileTerm();
    while (strnlen(currentToken, 2) == 1 && strchr(binOps, *currentToken)) {
        eatTT(SYMBOL);
        compileTerm();
    }
    endRule(EXPRESSION);
}
void compileTerm() {
    /*
    integerConstant | stringConstant | keywordConstant | varName|
    varName ' [' expression ']' | subroutineCall | ' (' expression ') ' | unaryOp term

    subroutineCall:
        subroutineName ' (' expressionList ') ' I ( className| varName) '. ' subroutineName
    '('expressionList")"
    */
    startRule(TERM);
    if (maybeEatTT(INT_CONST) || maybeEatTT(STR_CONST)) {
    } else if (maybeEatOneOf((char **)kwConstants, 4)) {
    } else if (maybeEat("(")) {
        compileExpr();
        eat(")");
    } else if (strnlen(currentToken, 2) == 1 && strchr(unaryOps, *currentToken)) {
        compileTerm();
    } else if (maybeEatTT(IDENTIFIER)) {
        if (maybeEat("[")) {
            compileExpr();
            eat("]");
        } else if (cmpTok(".") || cmpTok("(")) {
            compileSubCall();
        }
    }
    endRule(TERM);
}

void compileSubCall() {
    /*
     subroutineCall:
        subroutineName ' (' expressionList ') ' I ( className| varName) '. ' subroutineName
    '('expressionList")"
    */
    if (maybeEat("."))
        eatTT(IDENTIFIER);

    eat("(");
    compileExprList();
    eat(")");
}

void compileExprList() {
    /*
    (expression ("', ' expression)* )?
    */
    startRule(EXPRESSION_LIST);
    if (cmpTok(")")) {
        endRule(EXPRESSION_LIST);
        return;
    }
    compileExpr();
    while (maybeEat(","))
        compileExpr();
    endRule(EXPRESSION_LIST);
}

void advanceTokenizer() {
    if (hasMoreTokens()) {
        advance();
        currentToken = tokenVal();
        currentTokenType = tokenType();
    } else {
        printf("WARNING: Tried to advance tokenizer with no more tokens left.\n");
    }
}

void eatOneOf(char **vals, size_t len) {
    for (int i = 0; i < len; i++) {
        if (strncmp(currentToken, vals[i], MAX_TOKEN_SIZE) == 0) {
            writeToken();
            advanceTokenizer();
            return;
        }
    }

    char *joined = join(vals, ", ", len);
    char **ret = malloc(sizeof(char **));
    asprintf(ret, "Expected Token(s): %s", joined);
    error(*ret);
}

bool maybeEatOneOf(char **vals, size_t len) {
    for (int i = 0; i < len; i++) {
        if (strncmp(currentToken, vals[i], MAX_TOKEN_SIZE) == 0) {
            writeToken();
            advanceTokenizer();
            return true;
        }
    }
    return false;
}
void eatTT(TokenType tt) {
    if (currentTokenType == tt) {
        writeToken();
        advanceTokenizer();
        return;
    }
    char **ret = malloc(sizeof(char **));
    asprintf(ret, "Expected Token Type: %s", ttStr(currentTokenType));
    error(*ret);
}

// int, char, boolean or className (identifier)
void eatType() {
    int i = strncmp(currentToken, kwStr(INT), 3);
    int c = strncmp(currentToken, kwStr(CHAR), 4);
    int b = strncmp(currentToken, kwStr(BOOLEAN), 7);
    if (!i || !c || !b || currentTokenType == IDENTIFIER) {
        writeToken();
        advanceTokenizer();
        return;
    }

    error("Expected Type (int, char, boolean or className (identifier))");
}

bool maybeEat(char *expected) {
    if (!strncmp(currentToken, expected, strlen(expected))) {
        writeToken();
        advanceTokenizer();
        return true;
    }

    return false;
}

bool maybeEatTT(TokenType tt) {
    if (currentTokenType == tt) {
        writeToken();
        advanceTokenizer();
        return true;
    }
    return false;
}
void eat(char *str) {
    if (!maybeEat(str)) {
        char **ret = malloc(sizeof(char **));
        asprintf(ret, "Expected '%s'", str);
        error(*ret);
    }
}

void eatReturnType() {
    if (!maybeEat(kwStr(VOID)))
        eatType();
}
void writeToken() {
    char *tt = ttStr(currentTokenType);
    char *val = tokenVal();
    val = toXml(currentToken);
    fprintf(out, "<%s> %s </%s>\n", tt, val, tt);
}

void startRule(GrammarRule rule) {
    fprintf(out, "<%s>\n", ruleStr(rule));
}

void endRule(GrammarRule rule) {
    fprintf(out, "</%s>\n", ruleStr(rule));
}
bool cmpTok(char *s) {
    return !strncmp(currentToken, s, MAX_TOKEN_SIZE);
}
void error(char *msg) {
    printf("ERROR: %s\n", msg);
    printf("Current Token: %s\n", currentToken);
    printf("Current Line: %s\n", getLine());
    exit(1);
}

char *join(char **vals, char *sep, size_t len) {
    size_t seplen = strlen(sep);
    char *joined = calloc(1, strlen(vals[0]) + 1);
    strcpy(joined, vals[0]);
    for (int i = 1; i < len; i++) {
        size_t joinedlen = strlen(joined);
        size_t currValLen = strlen(vals[i]);
        joined = realloc(joined, joinedlen + seplen + currValLen + 1);
        strlcat(joined, sep, joinedlen + seplen + 1);
        strlcat(joined, vals[i], joinedlen + seplen + currValLen + 1);
    }

    return joined;
}